ql/src/java/org/apache/hadoop/hive/ql/lockmgr/HiveTxnManager.java

```diff
@@ -186,6 +186,18 @@ void replTableWriteIdState(String validWriteIdList, String dbName, String tableN
    */
   ValidTxnList getValidTxns() throws LockException;
 
+ /**
+  * Get the transactions that are currently valid.  The resulting
+  * {@link ValidTxnList} object can be passed as string to the processing
+  * tasks for use in the reading the data.  This call should be made once up
+  * front by the planner and should never be called on the backend,
+  * as this will violate the isolation level semantics.
+  * @return list of valid transactions.
+  * @param  txnTypes list of transaction types that should be excluded.
+  * @throws LockException
+  */
+  ValidTxnList getValidTxns(List<TxnType> txnTypes) throws LockException;
```

Comment:
txnTypes is set of 5 enum values. Its not a free flowing list. So filter/regex is not required. Same interface can be used to exclude any txntype by passing in that list.

Reply:
Currently txnTypes is set of 5 enum values, but that could change. Also, imagine if someone wants to get a single type, he will have to pass n-1 type in exclude list. Since it is exposed as HMS api, it could be used outside of replication cases as well. Since these are public interfaces, it is important. with this interface in place, if someone wants to add additional api to get txn with matching types, how will that extension look like?

Reply:
Using a regex is a overkill in this scenario. If we introduce a new txn type tomorrow and that matches this regex pattern, we will need to go an update at every place where that is defined.